src/REP15.sol

```diff
@@ -0,0 +1,466 @@
+  using REP15Utils for REP15Utils.TokenContext;
+
+  uint64 internal immutable _maxDetachingDuration;
+
+  mapping(uint256 tokenId => REP15Utils.Delegation) private _delegations;
+
+  mapping(bytes32 ctxHash => REP15Utils.Context) private _contexts;
+
+  mapping(uint256 tokenId => mapping(bytes32 ctxHash => REP15Utils.TokenContext)) private _tokenContext;
+
+  mapping(uint256 tokenId => bytes32[] ctxHashes) internal _attachedContexts;
+
+  mapping(uint256 tokenId => mapping(bytes32 ctxHash => uint256 index)) internal _attachedContextsIndex;
+
+  constructor(uint64 maxDetachingDurationSeconds) {
+    _maxDetachingDuration = maxDetachingDurationSeconds;
+  }
+
+  /**
+   * @inheritdoc IERC165
+   */
+  function supportsInterface(bytes4 interfaceId) public view virtual override(ERC721, IERC165) returns (bool) {
+    return interfaceId == type(IREP15).interfaceId || super.supportsInterface(interfaceId);
+  }
+
+  /**
+   * @inheritdoc IREP15
+   */
+  function startDelegateOwnership(uint256 tokenId, address delegatee, uint64 until) external virtual {
+    address owner = _requireOwned(tokenId);
+
+    if (delegatee == owner || delegatee == address(0)) revert REP15InvalidDelegatee(delegatee);
+
+    if (until <= block.timestamp) revert REP15InvalidDelegationExpiration(until);
+
+    REP15Utils.Delegation storage delegation = _delegations[tokenId];
+
+    if (delegation.isActive()) revert REP15AlreadyDelegatedOwnership(tokenId, delegation.delegatee, delegation.until);
+
+    ERC721._checkAuthorized(owner, _msgSender(), tokenId);
+
+    delegation.delegatee = delegatee;
+    delegation.until = until;
+
+    emit OwnershipDelegationStarted(tokenId, delegatee, until);
+  }
+
+  /**
+   * @inheritdoc IREP15
+   */
+  function acceptOwnershipDelegation(uint256 tokenId) external virtual {
+    REP15Utils.Delegation storage delegation = _delegations[tokenId];
+    address delegatee = delegation.delegatee;
+    uint64 until = delegation.until;
+
+    if (!delegation.isPending()) revert REP15NonexistentPendingOwnershipDelegation(tokenId);
+
+    _checkAuthorizedDelegatee(delegatee, _msgSender(), tokenId);
+
+    delegation.delegated = true;
+
+    emit OwnershipDelegationAccepted(tokenId, delegatee, until);
+  }
+
+  /**
+   * @inheritdoc IREP15
+   */
+  function stopOwnershipDelegation(uint256 tokenId) external virtual {
+    REP15Utils.Delegation storage delegation = _delegations[tokenId];
+    address delegatee = delegation.delegatee;
+
+    if (!delegation.isActive()) revert REP15InactiveOwnershipDelegation(tokenId);
+
+    _checkAuthorizedDelegatee(delegatee, _msgSender(), tokenId);
+
+    delete _delegations[tokenId];
+
+    emit OwnershipDelegationStopped(tokenId, delegatee);
+  }
+
+  /**
+   * @inheritdoc IREP15
+   */
+  function createContext(address controller, uint64 detachingDuration, bytes calldata ctxMsg)
+    external
+    virtual
+    returns (bytes32 ctxHash)
+  {
+    ctxHash = keccak256(abi.encodePacked(_msgSender(), ctxMsg));
```

Comment:
```suggestion
    ctxHash = keccak256(abi.encode(_msgSender(), keccak256(abi.encodePacked(ctxMsg))));
```

Reply:
please provide the PoC of this attack

Reply:
For hashing, we prefer using abi.encode in general.
See C#12 in: https://github.com/transmissions11/solcurity

Reply:
no attack, peace 🕊️

Reply:
how about keccak256(ctxMsg)?

Reply:
`keccak256(bytes(ctxMsg))` is the same as `keccak256(abi.encodePacked(ctxMsg))`

Reply:
are you ok with `keccak256(abi.encode(_msgSender(), ctxMsg))`?